nemo_curator/stages/text/io/writer/megatron_tokenizer.py

```diff
@@ -0,0 +1,222 @@
+    fields: list[str] | None = field(default=None, init=False, repr=False)
+
+    name: str = "megatron_tokenizer_writer"
+    file_extension: list[str] = field(default_factory=lambda: FILETYPE_TO_DEFAULT_EXTENSIONS["megatron"])
+
+    def __post_init__(self):
+        if self.model_identifier is None:
+            msg = "model_identifier is required and must be provided"
+            raise ValueError(msg)
+        super().__post_init__()
+
+    def setup_on_node(self, _node_info: NodeInfo | None = None, _worker_metadata: WorkerMetadata = None) -> None:
+        try:
+            snapshot_download(
+                repo_id=self.model_identifier,
+                cache_dir=self.cache_dir,
+                token=self.hf_token,
+            )
+        except Exception as e:
+            msg = f"Failed to download {self.model_identifier}"
+            raise RuntimeError(msg) from e
+
+    def setup(self, _worker_metadata: WorkerMetadata | None = None) -> None:
+        # Load the tokenizer
+        self.tokenizer = AutoTokenizer.from_pretrained(
+            self.model_identifier,
+            cache_dir=self.cache_dir,
+        )
+
+    def process(self, task: DocumentBatch) -> FileGroupTask:
+        sequence_lengths: list[int] = []
+        # Get source files from metadata for deterministic naming
+        if source_files := task._metadata.get("source_files"):
+            filename = writer_utils.get_deterministic_hash(source_files, task.task_id)
+        else:
+            logger.warning("The task does not have source_files in metadata, using UUID for base filename")
+            filename = uuid.uuid4().hex
+
+        file_prefix = self.fs.sep.join([self._fs_path, filename])
+        for file_extension in self.file_extension:
+            file_path = file_prefix + file_extension
+            if self.fs.exists(file_path):
+                logger.debug(f"File {file_path} already exists, overwriting it")
+
+        self.token_size = (
+            -1
+            if self.tokenizer.vocab_size is None
+            else (4 if self.tokenizer.vocab_size > np.iinfo(np.uint16).max + 1 else 2)
+        )
+        if self.token_size == -1:
+            logger.warning("tokenizer.vocab_size is not set, assuming 4 bytes per token (vocab_size > 65536)")
+            self.token_size = 4
+        self.token_dtype = np.int32 if self.token_size == 4 else np.uint16  # noqa: PLR2004
+        self.token_dtype_code = (
+            4 if self.token_size == 4 else 8  # noqa: PLR2004
+        )  # NOTE(asolergi-nv): Megatron needs this dtype code in the .idx file | https://github.com/NVIDIA/Megatron-LM/blob/64cbae55ac85cd73fbadbc3c0d715c8123c5e13b/megatron/core/datasets/indexed_dataset.py#L41
+
+        self.eod_token_id = self.tokenizer.eos_token_id if self.tokenizer.eos_token_id is not None else -1
+        if self.eod_token_id == -1:
+            logger.warning("tokenizer.eos_token_id is not set, disabling append_eod")
+            self.append_eod = False
```

Comment:
I would consider just using `token_size`, `token_dtype`, `token_dtype_code`, and `eod_token_id`, instead of `self.token_size`, etc., then passing them into the relevant functions, since this snippet is in `process`. Seems safer to me, LMK what you think.

And `bin_file` instead of `self.bin_file` too.